File: test/verification/test_design_rule_violations.cpp
/*
 * Copyright (c) 2018 - 2023 Marcel Walter
 * Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
 * All rights reserved.
 *
 * SPDX-License-Identifier: MIT
 *
 * Licensed under the MIT License
 */

/**
 * @file
 * @brief Tests for `fiction/verification/design_rule_violations.hpp`.
 * @author Marcel Walter (marcelwa)
 */

#include <catch2/catch_test_macros.hpp>

#include "utils/blueprints/layout_blueprints.hpp"

#include <fiction/types.hpp>
#include <fiction/verification/design_rule_violations.hpp>

#include <sstream>

using namespace fiction;
using namespace fiction::verification;

template <typename Lyt>
gate_level_drv_stats get_drvs(const Lyt& lyt)
{
    gate_level_drv_params ps{};
    gate_level_drv_stats  st{};

    // suppress standard output
    std::stringstream ss{};
    ps.out = &ss;

    gate_level_drvs(lyt, ps, &st);

    return st;
}

template <typename Lyt>
void check_for_drvs(const Lyt& lyt, const std::size_t num_drvs, const std::size_t num_warnings)
{
    const auto st = get_drvs(lyt);

    CHECK(st.drvs == num_drvs);
    CHECK(st.warnings == num_warnings);
}

TEST_CASE("Intact layouts", "[drv]")
{
    // empty layouts
    check_for_drvs(cart_gate_clk_lyt{}, 0, 0);
    check_for_drvs(hex_even_col_gate_clk_lyt{}, 0, 0);
    check_for_drvs(hex_odd_col_gate_clk_lyt{}, 0, 0);
    check_for_drvs(hex_even_row_gate_clk_lyt{}, 0, 0);
    check_for_drvs(hex_odd_row_gate_clk_lyt{}, 0, 0);

    // Cartesian gate layouts
    check_for_drvs(blueprints::and_or_gate_layout<cart_gate_clk_lyt>(), 0, 0);
    check_for_drvs(blueprints::and_not_gate_layout<cart_gate_clk_lyt>(), 0, 0);
    check_for_drvs(blueprints::or_not_gate_layout<cart_gate_clk_lyt>(), 0, 0);
    check_for_drvs(blueprints::crossing_layout<cart_gate_clk_lyt>(), 0, 0);
    check_for_drvs(blueprints::fanout_layout<cart_gate_clk_lyt>(), 0, 0);
    check_for_drvs(blueprints::unbalanced_and_layout<cart_gate_clk_lyt>(), 0, 0);
    check_for_drvs(blueprints::se_gate_layout<cart_gate_clk_lyt>(), 0, 0);
}

TEST_CASE("Warnings", "[drv]")
{
    // PI tile (1,1) is not at a border
    check_for_drvs(blueprints::xor_maj_gate_layout<cart_gate_clk_lyt>(), 0, 1);
}

TEST_CASE("DRVs", "[drv]")
{
    check_for_drvs(blueprints::non_structural_all_function_gate_layout<cart_gate_clk_lyt>(), 50, 1);
}

File: test/python/verification/test_design_rule_violations.py
# Copyright (c) 2018 - 2023 Marcel Walter
# Copyright (c) 2023 - present Chair for Design Automation, Technical University of Munich
# All rights reserved.
#
# SPDX-License-Identifier: MIT
#
# Licensed under the MIT License

from __future__ import annotations

from mnt.pyfiction.layouts import cartesian_gate_layout
from mnt.pyfiction.layouts.coords import offset_coordinate
from mnt.pyfiction.physical_design import color_routing
from mnt.pyfiction.verification import gate_level_drvs


def test_drvs():
    # Create empty layout
    layout = cartesian_gate_layout(offset_coordinate(2, 5, 0), "2DDWave")

    # Create 2:1 MUX

    # Inputs
    layout.create_pi("x1", offset_coordinate(0, 3))
    layout.create_pi("x2", offset_coordinate(0, 0))
    layout.create_pi("x3", offset_coordinate(2, 0))

    # Wires
    layout.create_buf(layout.make_signal(layout.get_node(offset_coordinate(0, 0, 0))), offset_coordinate(0, 1))
    layout.create_buf(layout.make_signal(layout.get_node(offset_coordinate(0, 1, 0))), offset_coordinate(1, 1))

    # NOT
    layout.create_not(layout.make_signal(layout.get_node(offset_coordinate(0, 1, 0))), offset_coordinate(0, 2))

    # Wires
    layout.create_buf(layout.make_signal(layout.get_node(offset_coordinate(0, 2, 0))), offset_coordinate(1, 2))
    # AND
    layout.create_and(
        layout.make_signal(layout.get_node(offset_coordinate(0, 3, 0))),
        layout.make_signal(layout.get_node(offset_coordinate(1, 2, 0))),
        offset_coordinate(1, 3),
    )

    # AND
    layout.create_and(
        layout.make_signal(layout.get_node(offset_coordinate(1, 1, 0))),
        layout.make_signal(layout.get_node(offset_coordinate(2, 0, 0))),
        offset_coordinate(2, 1),
    )

    # Wires
    layout.create_buf(layout.make_signal(layout.get_node(offset_coordinate(2, 1, 0))), offset_coordinate(2, 2))

    # OR
    layout.create_or(
        layout.make_signal(layout.get_node(offset_coordinate(1, 3, 0))),
        layout.make_signal(layout.get_node(offset_coordinate(2, 2, 0))),
        offset_coordinate(2, 3),
    )

    # Outputs
    layout.create_po(layout.make_signal(layout.get_node(offset_coordinate(2, 3, 0))), "f1", offset_coordinate(2, 4))

    layout.move_node(layout.get_node(offset_coordinate(2, 4)), offset_coordinate(2, 5))

    color_routing(layout, [(offset_coordinate(2, 3), offset_coordinate(2, 5))])

    warnings, drvs = gate_level_drvs(layout)

    assert warnings == 0
    assert drvs == 0
